src/metrics/dns.rs

```diff
@@ -0,0 +1,278 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::identity::Identity;
+use crate::metrics::{DefaultedUnknown, Recorder};
+use crate::state::workload::address::Address;
+use crate::state::workload::Workload;
+use prometheus_client::encoding::EncodeLabelSet;
+use prometheus_client::metrics::counter::Counter;
+use prometheus_client::metrics::family::Family;
+use prometheus_client::metrics::histogram::Histogram;
+use prometheus_client::registry::Registry;
+use std::time::Duration;
+use trust_dns_server::server::Request;
+
+pub struct Metrics {
+    pub requests: Family<DnsLabels, Counter>,
+    pub forwarded_requests: Family<DnsLabels, Counter>,
+    pub forwarded_failures: Family<DnsLabels, Counter>,
+    pub forwarded_duration: Family<DnsLabels, Histogram>,
+}
+
+impl Metrics {
+    pub fn new(registry: &mut Registry) -> Self {
+        let requests = Family::default();
+        registry.register(
+            "dns_requests_total",
+            "Total number of DNS requests.",
+            requests.clone(),
+        );
+
+        let forwarded_requests = Family::default();
+        registry.register(
+            "dns_upstream_requests_total",
+            "Total number of DNS requests forwarded to upstream.",
+            forwarded_requests.clone(),
+        );
+
+        let forwarded_failures = Family::default();
+        registry.register(
+            "dns_upstream_failures_total",
+            "Total number of DNS requests forwarded to upstream.",
+            forwarded_failures.clone(),
+        );
+
+        let forwarded_duration = Family::<DnsLabels, Histogram>::new_with_constructor(|| {
+            Histogram::new(vec![0.005f64, 0.001, 0.01, 0.1, 1.0, 5.0].into_iter())
+        });
+        registry.register(
+            "dns_upstream_request_duration_seconds",
+            "Total time in seconds Istio takes to get DNS response from upstream.",
+            forwarded_duration.clone(),
+        );
+
+        Self {
+            requests,
+            forwarded_requests,
+            forwarded_failures,
+            forwarded_duration,
+        }
+    }
+}
+
+#[derive(Clone, Hash, Debug, PartialEq, Eq, EncodeLabelSet)]
+pub struct DnsLabels {
```

Comment:
This seems like a LOT of labels. I am not sure this much is appropriate for Prometheus. Source IP is unbounded cardinality for example

Reply:
I am a bit torn. I see the value but also the cost. I will think about it a bit more (and welcome other opinions here)